what4/src/What4/Protocol/VerilogWriter/Backend.hs

```diff
@@ -0,0 +1,398 @@
+  binop Eq x' y'
+
+bvarIdentifier :: ExprBoundVar t tp -> Identifier
+bvarIdentifier x = show (bvarName x)
+
+nonceAppExprVerilogExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  NonceAppExpr n tp ->
+  VerilogM sym n (IExp tp)
+nonceAppExprVerilogExpr nae =
+  case nonceExprApp nae of
+    Forall _ _ -> doNotSupportError "universal quantification"
+    Exists _ _ -> doNotSupportError "existential quantification"
+    ArrayFromFn _ -> doNotSupportError "arrays"
+    MapOverArrays _ _ _ -> doNotSupportError "arrays"
+    ArrayTrueOnEntries _ _ -> doNotSupportError "arrays"
+    FnApp f Empty -> do
+      name <- addFreshInput tp base
+      return $ Ident tp name
+        where
+          tp = symFnReturnType f
+          base = show (symFnName f)
+    -- TODO: inline defined functions?
+    -- TODO: implement uninterpreted functions as uninterpreted functions
+    FnApp _ _ -> doNotSupportError "named function applications"
+    Annotation _ _ e -> exprToVerilogExpr e
+
+-- | As of now, all symbolic functions are just turned into verilog functions of
+-- the same name, but it may be the case that some pre-defined uninterpreted
+-- functions in What4 should be transformed into other operations in verilog. OR
+-- maybe symbolic functions should be turned into module calls?
+mkVerilogFn :: ExprSymFn t (Expr t) args ret -> String
+mkVerilogFn f = show (symFnName f)
+
+boolMapToExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  Bool ->
+  Bool ->
+  Binop WT.BaseBoolType WT.BaseBoolType ->
+  BMap.BoolMap (Expr n) ->
+  VerilogM sym n (IExp WT.BaseBoolType)
+boolMapToExpr u du op es =
+  let pol (x,Positive) = exprToVerilogExpr x
+      pol (x,Negative) = unop Not =<< exprToVerilogExpr x
+  in
+  case BMap.viewBoolMap es of
+    BMap.BoolMapUnit -> litBool u
+    BMap.BoolMapDualUnit -> litBool du
+    BMap.BoolMapTerms (t:|[]) -> pol t
+    BMap.BoolMapTerms (t:|ts) -> do
+      t' <- pol t
+      ts' <- mapM pol ts
+      foldM (binop op) t' ts'
+
+leqSubPos :: (1 <= m, 1 <= n, n+1 <= m) => NatRepr m -> NatRepr n -> LeqProof 1 (m - n)
+leqSubPos mr nr =
+  case (plusComm nr one, plusMinusCancel one nr) of
+    (Refl, Refl) ->
+      leqSub2 (leqProof (nr `addNat` one) mr) (leqProof nr nr)
+  where one = knownNat :: NatRepr 1
+
+leqSuccLeft :: (n + 1 <= m) => p m -> NatRepr n -> LeqProof n m
+leqSuccLeft mr nr =
+  case (plusComm nr one, addPrefixIsLeq nr one) of
+    (Refl, LeqProof) ->
+      leqTrans (addIsLeq nr one) (leqProof (nr `addNat` one) mr)
+  where one = knownNat :: NatRepr 1
+
+appExprVerilogExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  AppExpr n tp ->
+  VerilogM sym n (IExp tp)
+appExprVerilogExpr = appVerilogExpr . appExprApp
+
+appVerilogExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  App (Expr n) tp ->
+  VerilogM sym n (IExp tp)
+appVerilogExpr app =
+  case app of
+    -- Generic operations
+    BaseIte _ _ b etrue efalse -> do
+      b' <- exprToVerilogExpr b
```

Comment:
These calls should not be to `exprToVerilogExpr`, as they will each create fresh cache values.  Instead a single cache should be used for the entire evaluation.

Reply:
Oh, good catch!

Reply:
Actually, I just took a closer look... I was mistaken about this, the cache is actually held in the monad state.  Let me take another look to be sure I've understood what's happening.

Reply:
Yeah, whoops, I didn't read things closely enough.  It looks like the cache is being properly piped around via the monad being used, and is only created once in the right place.

Reply:
Oh, yeah, the `cache <- vsExpCache <$> get` line isn't creating a new cache.

Reply:
I'm still going to get rid of the special function for equality between expressions. I don't think we need it.